src/navier.f90

```diff
@@ -1280,5 +1289,246 @@ subroutine tbl_flrt (ux1,uy1,uz1)
+                    ux(i,j,k)=ux(i,j,k)+(ub_constant-qm)
+                else
+                    !Cancel solid zone (rai <= r <= rao)
+                    !and buffer zone (r > rao)
+                    ux(i,j,k)=zero
+                    uy(i,j,k)=zero
+                    uz(i,j,k)=zero
+                endif
+            enddo
+        enddo
+    enddo
+
+    !Check new bulk velocity
+    if (mod(itime, ilist)==0) then
+        call pipe_volume_avg(ux,qm,ep)
+        if (nrank==0) print *,'    Bulk velocity  after',qm
+    endif
+    !
+    return
+    !
+  end subroutine pipe_bulk_u
+  !********************************************************************
+  !
+  subroutine pipe_bulk_phi(phi,ux,ep,is,phib_constant)
+  !
+  !********************************************************************
+
+    use decomp_2d
+    use decomp_2d_poisson
+    use variables
+    use param
+    use ibm_param, only: rai
+    use MPI
+
+    implicit none
+    !INPUTS
+    real(mytype),intent(inout),dimension(xsize(1),xsize(2),xsize(3))  :: phi,ux
+    real(mytype),intent(in   ),dimension(xsize(1),xsize(2),xsize(3))  :: ep
+    real(mytype),intent(in   )                                        :: phib_constant !bulk temperature value
+    !LOCALS
+    real(mytype)                                                      :: qv,qm !volumetric averaged values
+    real(mytype)                                                      :: ym,zm,yc,zc,r
+    real(mytype)                                                      :: ncount
+    real(mytype)                                                      :: phi_out
+    integer                                                           :: is,j,i,k,code
+    character(len=30)                                                 :: filename
+    integer, allocatable, save, dimension(:)                          :: local_io_unit
+
+    if (iscalar.eq.0) return
+
+255 format('Tb_',I2.2,'.dat')
+256 format(' Scalar:                       #',I2)
+
+    ! Safety check
+    if (is<1.or.is>numscalar) return
+
+    if (.not.allocated(local_io_unit)) then
+        allocate(local_io_unit(numscalar))
+        local_io_unit(:)=-1
+    endif
+
+    if (local_io_unit(is).eq.-1.and.nrank.eq.0) then
+        write(filename,255) is
+        open(newunit=local_io_unit(is),file=filename,status='unknown')
+    endif
+
+    yc = yly / two
+    zc = zlz / two
+    phi_out =zero !for reconstruction smoothness
+
+    !--------------------------- Bulk Temperature ---------------------------
+    !                  with corrected streamwise velocity
+    call pipe_volume_avg(ux*phi,qm,ep)
+    call pipe_volume_avg(ux*ux ,qv,ep)
+    if (nrank.eq.0) then
+        if (mod(itime, ilist)==0) write(*,256) is
+        if (mod(itime, ilist)==0) print *,'         Bulk phi before',qm
+        write(local_io_unit(is),*) real((itime-1)*dt,mytype),(phib_constant-qm)/(qv)
+    endif
+
+    !Correction
+    do j=1,xsize(2)
```

Comment:
Loop order ?

Reply:
Solved in 25d46cc43400701dbc33aadb5d8c25638838e334.